rocAL/rocAL/source/pipeline/tensor.cpp

```diff
@@ -0,0 +1,374 @@
+
+        for (unsigned i = 0; i < info().batch_size(); i++)
+        {
+            if (width[i] > max_width)
+            {
+                ERR("Given ROI width is larger than buffer width for tensor[" + TOSTR(i) + "] " + TOSTR(width[i]) + " > " + TOSTR(max_width))
+                _info.get_roi()->at(i).x2 = max_width;
+            }
+            else
+            {
+                _info.get_roi()->at(i).x2 = width[i];
+            }
+            if (height[i] > max_height)
+            {
+                ERR("Given ROI height is larger than buffer with for tensor[" + TOSTR(i) + "] " + TOSTR(height[i]) + " > " + TOSTR(max_height))
+                _info.get_roi()->at(i).y2 = max_height;
+            }
+            else
+            {
+                _info.get_roi()->at(i).y2 = height[i];
+            }
+        }
+    }
+}
+
+rocALTensor::~rocALTensor()
+{
+    _mem_handle = nullptr;
+    if(_vx_handle)
+        vxReleaseTensor(&_vx_handle);
+}
+
+//! Converts the Rocal data_type to OpenVX
+vx_enum interpret_tensor_data_type(RocalTensorDataType data_type)
+{
+    switch (data_type)
+    {
+        case RocalTensorDataType::FP32:
+            return VX_TYPE_FLOAT32;
+        case RocalTensorDataType::FP16:
+            return VX_TYPE_FLOAT16;
+        case RocalTensorDataType::UINT8:
+            return VX_TYPE_UINT8;
+        default:
+            THROW("Unsupported Tensor type " + TOSTR(data_type))
+    }
+}
+
+rocALTensor::rocALTensor(const rocALTensorInfo &tensor_info) : _info(tensor_info)
+{
+    _info._type = rocALTensorInfo::Type::UNKNOWN;
+    _mem_handle = nullptr;
+}
+
+int rocALTensor::create_virtual(vx_context context, vx_graph graph)
+{
+    if (_vx_handle)
+        return -1;
+
+    _context = context;
+
+    unsigned num_of_dims = _info.num_of_dims();
+    vx_size dims[num_of_dims];
+    for(unsigned i = 0; i < num_of_dims; i++)
+    {
+        dims[i] = _info.dims().at(i);
+    }
+    _vx_handle = vxCreateVirtualTensor(graph, num_of_dims, dims, interpret_tensor_data_type(_info.data_type()), 0);
+    vx_status status;
+    if ((status = vxGetStatus((vx_reference)_vx_handle)) != VX_SUCCESS)
+        THROW("Error: vxCreateVirtualTensor(input:[" + TOSTR(_info.max_dims().at(0)) + "W" + TOSTR(_info.max_dims().at(1)) + "H" + "]): failed " + TOSTR(status))
+
+    _info._type = rocALTensorInfo::Type::VIRTUAL;
+    return 0;
+}
+
+int rocALTensor::create_from_handle(vx_context context)
+{
+    if (_vx_handle)
+    {
+        WRN("Tensor object create method is already called ")
+        return -1;
+    }
+
+    _context = context;
+    vx_enum tensor_data_type = interpret_tensor_data_type(_info.data_type());
+    unsigned num_of_dims = _info.num_of_dims();
+    vx_size stride[num_of_dims];
+    void *ptr[1] = {nullptr};
+
+    stride[0] = tensor_data_size(_info.data_type());
+    for(unsigned i = 1; i < num_of_dims; i++)
+        stride[i] = stride[i - 1] * _info.dims().at(i - 1);
```

Comment:
Can we eliminate the stride calculation and query it from vx_tensor?

Reply:
In order to create the vx_tensor we would need to pass the strides.